openmc/tallies.py

```diff
@@ -1620,108 +1660,129 @@ def _align_tally_data(self, other):
+            other._std_dev = np.repeat(other.std_dev, filter.num_bins, axis=0)
+            other.add_filter(filter)
+
+        # Add filters present in other but not in self to self
+        for filter in self_missing_filters:
+            filter = copy.deepcopy(filter)
+            self._mean = np.repeat(self.mean, filter.num_bins, axis=0)
+            self._std_dev = np.repeat(self.std_dev, filter.num_bins, axis=0)
+            self.add_filter(filter)
+
+        # Align other filters with self filters
+        for i, filter in enumerate(self.filters):
+            other_index = other.filters.index(filter)
+
+            # If necessary, swap other filter
+            if other_index != i:
+                other._swap_filters(filter, other.filters[i])
+
+        # Repeat and tile the data by nuclide in preparation for performing
+        # the tensor product across nuclides.
+        if nuclide_product == 'tensor':
+            self._mean = np.repeat(self.mean, other.num_nuclides, axis=1)
+            self._std_dev = np.repeat(self.std_dev, other.num_nuclides, axis=1)
+            other._mean = np.tile(other.mean, (1, self.num_nuclides, 1))
+            other._std_dev = np.tile(other.std_dev, (1, self.num_nuclides, 1))
+
+        # Add nuclides to each tally such that each tally contains the complete
+        # set of nuclides necessary to perform an entrywise product. New nuclides
+        # added to a tally will have all their scores set to zero.
+        else:
+
+            # Get the set of nuclides that each tally is missing
+            other_missing_nuclides = set(self.nuclides).difference(set(other.nuclides))
+            self_missing_nuclides = set(other.nuclides).difference(set(self.nuclides))
+
+            # Add nuclides present in self but not in other to other
+            for nuclide in other_missing_nuclides:
+                other._mean = np.insert(other.mean, other.num_nuclides, 0, axis=1)
+                other._std_dev = np.insert(other.std_dev, other.num_nuclides, 0, axis=1)
+                other.add_nuclide(nuclide)
+
+            # Add nuclides present in other but not in self to self
+            for nuclide in self_missing_nuclides:
+                self._mean = np.insert(self.mean, self.num_nuclides, 0, axis=1)
+                self._std_dev = np.insert(self.std_dev, self.num_nuclides, 0, axis=1)
+                self.add_nuclide(nuclide)
+
+            # Align other nuclides with self nuclides
+            for i, nuclide in enumerate(self.nuclides):
+                other_index = other.get_nuclide_index(nuclide)
+
+                # If necessary, swap other nuclide
+                if other_index != i:
+                    other._swap_nuclides(nuclide, other.nuclides[i])
+
+        # Repeat and tile the data by score in preparation for performing
+        # the tensor product across scores.
+        if score_product == 'tensor':
+            self._mean = np.repeat(self.mean, other.num_score_bins, axis=2)
```

Comment:
As I kind of alluded to before, `num_scores` is equivalent to `num_score_bins` for the case where duplicate scores are not present. If they are present, they will be different.